ingest/cas.py

```diff
@@ -3,24 +3,204 @@
+
+    return [adata, cas_ontology_aware_response]
+
+def make_compliant_for_scp(adata):
+    """Shim to make AnnData file compliant with SCP metadata schema
+
+    Only use for demo / development purposes.  This is generally commented out
+    upstream.
+    """
+    if 'biosample_id' not in adata.obs:
+        adata.obs['biosample_id'] = "sample-1"
+        adata.obs['donor_id'] = "donor-1"
+        adata.obs['species'] = "NCBITaxon_9606"
+        adata.obs['species__ontology_label'] = "Homo sapiens"
+        adata.obs['disease'] = "PATO_0000461"
+        adata.obs['disease__ontology_label'] = "normal"
+        adata.obs['organ'] = "UBERON_0000178"
+        adata.obs['organ__ontology_label'] = "blood"
+        adata.obs['library_preparation_protocol'] = "EFO_0030059"
+        adata.obs['library_preparation_protocol__ontology_label'] = "10x 3' v3"
+        adata.obs['sex'] = "female"
+    return adata
+
+# Stub for potential later expansion
+# def format_as_scp_metadatum(cas_response_path):
+#     with open(cas_response_filepath) as f:
+#         cas_response = json.loads(f.read())
+
+#     tsv_rows = []
+#     for cas_item in cas_response:
+#         cell = cas_item["query_cell_id"]
+#         first_match = cas_item["matches"][0]
+#         annotation_label = cas_item["matches"][0]
+
+def merge_cas_results(adata, cas_ontology_aware_response):
+    """Update AnnData with ontology-aware CAS results JSON
+    """
+    insert_cas_ontology_aware_response_into_adata(cas_ontology_aware_response, adata, cl)
+
+    # Comment out block below unless debugging
+    # DASH_SERVER_PORT = 8050
+    # with suppress_stderr():
+    #     CASCircularTreePlotUMAPDashApp(
+    #         adata,  # the AnnData file
+    #         cas_ontology_aware_response,  # CAS response
+    #         cluster_label_obs_column="cluster_label",  # (optional) The .obs column name containing cluster labels
+    #     ).run(port=DASH_SERVER_PORT, debug=False, jupyter_width="100%")
+
+    # TODO: Consider conditionally using this `single` method if (and only if)
+    # AnnData lacks "raw annotations" as returned from e.g. Scanpy or Seurat
+    # ("0", "1", "2", etc.)
+    # pp.compute_most_granular_top_k_calls_single(
+    #     adata=adata,
+    #     cl=cl,
+    #     min_acceptable_score=0.1,  # minimum acceptable score for a call
+    #     top_k=3,  # how many top calls to make?
+    #     obs_prefix="cell_type_cas"  # .obs column to write the top-k calls to
+    # )
+
+    # Use this only if source AnnData has clusterings
+    pp.compute_most_granular_top_k_calls_cluster(
+        adata=adata,
+        cl=cl,
+        min_acceptable_score=0.1,  # minimum acceptable score for a call
+        cluster_label_obs_column='cluster_label',  # .obs column containing cluster labels
+        top_k=3,  # how many top calls to make?
+        obs_prefix='cell_type_cas'  # .obs column to write the top-k calls to
+    )
+
+    return adata
+
+def trim_cas_adata(adata):
```

Comment:
As a study owner, I'd probably be interested in the score.... For SCP, we may want to take a look at scores and ask CAS-interested SCP users when/how they would use thresholding with scores to interpret CAS results.

Reply:
That makes sense!  The [CAS notebook](https://github.com/cellarium-ai/cellarium-cas/blob/main/notebooks/quickstart_tutorial.ipynb) has a visualization that enables users to see CAS cell type assignments in a tree plot.  It colors cell type nodes by score, has sliders to set score thresholds, etc.  If we ask users about incorporating scores into CAS-SCP views, that's one option to consider.